README.md
Embedded JSON database library Python 2.7/3.x binding
============================================================

Installation
---------------------------------

**Required tools/system libraries:**

* gcc
* **Python >= 3.x|2.7.x**
* EJDB C library **libejdb** ([from sources](https://github.com/Softmotions/ejdb#building) or as [debian packages](https://github.com/Softmotions/ejdb#ubuntu))

**Python binding published:** http://pypi.python.org/pypi/pyejdb

**(A) Installing using pip**

`pip` for python3 or python2 should be installed (`sudo apt-get install python3-pip` | `sudo apt-get install python-pip`)

```
umask 022

   sudo pip install pyejdb

Upgrading:
   sudo pip install pyejdb --upgrade
```

**(B) Installing directly from sources**

```
umask 022

git clone https://github.com/Softmotions/ejdb-python.git
cd ./ejdb-python
sudo python3 ./setup.py install
```


**(C) Installing on Ubuntu/Debian (only for Python3.x)**

```
sudo add-apt-repository ppa:adamansky/ejdb
sudo apt-get update
sudo apt-get install python3-ejdb
```


One snippet intro
---------------------------------

```python
import pyejdb
from datetime import datetime

#Open database
ejdb = pyejdb.EJDB("zoo", pyejdb.DEFAULT_OPEN_MODE | pyejdb.JBOTRUNC)

parrot1 = {
    "name": "Grenny",
    "type": "African Grey",
    "male": True,
    "age": 1,
    "birthdate": datetime.utcnow(),
    "likes": ["green color", "night", "toys"],
    "extra1": None
}
parrot2 = {
    "name": "Bounty",
    "type": "Cockatoo",
    "male": False,
    "age": 15,
    "birthdate": datetime.utcnow(),
    "likes": ["sugar cane"],
    "extra1": None
}
ejdb.save("parrots2", parrot1, parrot2)

with ejdb.find("parrots2", {"likes" : "toys"},
          hints={"$orderby" : [("name", 1)]}) as cur:
    print("found %s parrots" % len(cur))
    for p in cur:
        print("%s likes toys!" % p["name"])

ejdb.close()
```



debian/changelog
python-ejdb (1.0.16-ppa1) testing; urgency=low

  * EJDB 1.2.3

 -- Anton Adamansky <adamansky@gmail.com>  Fri, 20 Mar 2015 12:04:03 +0600

python-ejdb (1.0.15) testing; urgency=low

  * EJDB 1.1.14

 -- Anton Adamansky <adamansky@gmail.com>  Thu, 18 Jul 2013 18:00:55 +0700

python-ejdb (1.0.14) testing; urgency=low

  * Critical bugfix: error on saving docs.

 -- Anton Adamansky <adamansky@gmail.com>  Mon, 08 Jul 2013 11:43:54 +0700

python-ejdb (1.0.11-1) testing; urgency=low

  * Unified 2.7.x & 3.x python binding

 -- Anton Adamansky <adamansky@gmail.com>  Thu, 30 May 2013 18:03:41 +0700

python-ejdb (1.0.7-1) unstable; urgency=low

  * Initial release.

 -- Anton Adamansky <adamansky@gmail.com>  Fri, 15 Feb 2013 19:45:15 +0700


test/test_one.py
# -*- coding: utf8 -*-

# *************************************************************************************************
#  Python API for EJDB database library http://ejdb.org
#  Copyright (C) 2012-2013 Softmotions Ltd.
#
#  This file is part of EJDB.
#  EJDB is free software; you can redistribute it and/or modify it under the terms of
#  the GNU Lesser General Public License as published by the Free Software Foundation; either
#  version 2.1 of the License or any later version.  EJDB is distributed in the hope
#  that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
#  License for more details.
#  You should have received a copy of the GNU Lesser General Public License along with EJDB;
#  if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
#  Boston, MA 02111-1307 USA.
# *************************************************************************************************
from __future__ import with_statement
from __future__ import division
from __future__ import print_function

from datetime import datetime
import sys

PY3 = sys.version_info[0] == 3

import unittest
from pyejdb import bson
import pyejdb

if PY3:
    from io import StringIO as strio
else:
    from io import BytesIO as strio


class TestOne(unittest.TestCase):
    def __init__(self, *args, **kwargs):
        super(TestOne, self).__init__(*args, **kwargs)
        #super().__init__(*args, **kwargs)

    _ejdb = None

    @classmethod
    def setUpClass(cls):
        print("pyejdb version: %s" % pyejdb.version)
        print("libejdb_version: %s" % pyejdb.libejdb_version)
        cls._ejdb = pyejdb.EJDB("testdb", pyejdb.DEFAULT_OPEN_MODE | pyejdb.JBOTRUNC)

    def test(self):
        ejdb = TestOne._ejdb
        self.assertEqual(ejdb.isopen, True)
        doc = {"foo": "bar", "foo2": 2}
        ejdb.save("foocoll", doc)
        self.assertEqual(type(doc["_id"]).__name__, "str" if PY3 else "unicode")
        ldoc = ejdb.load("foocoll", doc["_id"])
        self.assertIsInstance(ldoc, dict)
        self.assertEqual(doc["_id"], ldoc["_id"])
        self.assertEqual(doc["foo"], ldoc["foo"])
        self.assertEqual(doc["foo2"], ldoc["foo2"])

        cur = ejdb.find("foocoll", {"foo": "bar"}, hints={"$fields": {"foo2": 0}})
        self.assertEqual(len(cur), 1)
        d = cur[0]
        self.assertTrue(d is not None)
        self.assertEqual(d["_id"], ldoc["_id"])

        with ejdb.find("foocoll") as cur2:
            d = cur2[0]
            self.assertTrue(d is not None)
            self.assertEqual(d["_id"], ldoc["_id"])

        self.assertEqual(ejdb.findOne("foocoll")["foo"], "bar")
        self.assertTrue(ejdb.findOne("foocoll2") is None)

        self.assertEqual(ejdb.count("foocoll"), 1)
        self.assertEqual(ejdb.count("foocoll2"), 0)

        ejdb.ensureStringIndex("foocoll", "foo")
        cur = ejdb.find("foocoll", {"foo": "bar"}, hints={"$fields": {"foo2": 0}})
        self.assertEqual(len(cur), 1)

        ejdb.remove("foocoll", doc["_id"])
        ldoc = ejdb.load("foocoll", doc["_id"])
        self.assertTrue(ldoc is None)
        ejdb.sync()

        ejdb.ensureCollection("ecoll1", records=90000, large=False)
        ejdb.dropCollection("ecoll1", prune=True)


    def test2(self):
        ejdb = TestOne._ejdb
        self.assertEqual(ejdb.isopen, True)
        parrot1 = {
            "name": "Grenny",
            "type": "African Grey",
            "male": True,
            "age": 1,
            "birthdate": datetime.utcnow(),
            "likes": ["green color", "night", "toys"],
            "extra1": None
        }

        parrot2 = {
            "name": "Bounty",
            "type": "Cockatoo",
            "male": False,
            "age": 15,
            "birthdate": datetime.utcnow(),
            "likes": ["sugar cane"],
            "extra1": None
        }
        ejdb.save("parrots", *[parrot1, None, parrot2])
        self.assertEqual(type(parrot1["_id"]).__name__, "str" if PY3 else "unicode")
        self.assertEqual(type(parrot2["_id"]).__name__, "str" if PY3 else "unicode")
        p2 = ejdb.load("parrots", parrot2["_id"])
        self.assertEqual(p2["_id"], parrot2["_id"])

        cur = ejdb.find("parrots")
        self.assertEqual(len(cur), 2)
        self.assertEqual(len(cur[1:]), 1)
        self.assertEqual(len(cur[2:]), 0)

        cur = ejdb.find("parrots",
                        {"name": bson.BSON_Regex(("(grenny|bounty)", "i"))},
                        hints={"$orderby": [("name", 1)]})
        self.assertEqual(len(cur), 2)
        self.assertEqual(cur[0]["name"], "Bounty")
        self.assertEqual(cur[0]["age"], 15)

        cur = ejdb.find("parrots", {}, {"name": "Grenny"}, {"name": "Bounty"},
                        hints={"$orderby": [("name", 1)]})
        self.assertEqual(len(cur), 2)

        cur = ejdb.find("parrots", {}, {"name": "Grenny"},
                        hints={"$orderby": [("name", 1)]})
        self.assertEqual(len(cur), 1)

        sally = {
            "name": "Sally",
            "mood": "Angry",
        }
        molly = {
            "name": "Molly",
            "mood": "Very angry",
            "secret": None
        }
        ejdb.save("birds", *[sally, molly])

        logbuf = strio()
        ejdb.find("birds", {"name": "Molly"}, log=logbuf)
        #print("LB=%s" % logbuf.getvalue())
        self.assertTrue(logbuf.getvalue().find("RUN FULLSCAN") != -1)

        ejdb.ensureStringIndex("birds", "name")

        logbuf = strio()
        ejdb.find("birds", {"name": "Molly"}, log=logbuf)
        self.assertTrue(logbuf.getvalue().find("MAIN IDX: 'sname'") != -1)
        self.assertTrue(logbuf.getvalue().find("RUN FULLSCAN") == -1)

        ##print("dbmeta=%s" % ejdb.dbmeta())
        bar = {
            "foo": "bar"
        }
        self.assertEqual(ejdb.isactivetx("bars"), False)
        ejdb.begintx("bars")
        self.assertEqual(ejdb.isactivetx("bars"), True)
        ejdb.save("bars", bar)
        self.assertTrue(bar["_id"] is not None)
        ejdb.abortx("bars")
        self.assertTrue(ejdb.load("bars", bar["_id"]) is None)

        ejdb.begintx("bars")
        ejdb.save("bars", bar)
        self.assertTrue(ejdb.load("bars", bar["_id"]) is not None)
        self.assertEqual(ejdb.isactivetx("bars"), True)
        ejdb.commitx("bars")
        self.assertEqual(ejdb.isactivetx("bars"), False)
        self.assertTrue(ejdb.load("bars", bar["_id"]) is not None)

        ejdb.update("upsertcoll",
                    {"foo": "bar", "$upsert": {"foo": "bar"}})
        self.assertTrue(ejdb.findOne("upsertcoll", {"foo": "bar"}) is not None)

        cmd = {
            "ping": {}
        }
        cmdret = ejdb.command(cmd)
        self.assertIsNotNone(cmdret)
        self.assertEqual(cmdret["log"], "pong")


    @classmethod
    def tearDownClass(cls):
        if cls._ejdb:
            cls._ejdb.close()
            cls._ejdb = None


if __name__ == '__main__':
    unittest.main()






src/pyejdb.c
/**************************************************************************************************
 *  Python API for EJDB database library http://ejdb.org
 *  Copyright (C) 2012-2013 Softmotions Ltd <info@softmotions.com>
 *
 *  This file is part of EJDB.
 *  EJDB is free software; you can redistribute it and/or modify it under the terms of
 *  the GNU Lesser General Public License as published by the Free Software Foundation; either
 *  version 2.1 of the License or any later version.  EJDB is distributed in the hope
 *  that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 *  License for more details.
 *  You should have received a copy of the GNU Lesser General Public License along with EJDB;
 *  if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 *  Boston, MA 02111-1307 USA.
 *************************************************************************************************/

#include "pyejdb.h"

/* PEJDB */
typedef struct {
    PyObject_HEAD
    EJDB *ejdb;
} PEJDB;

/* Cursor */
typedef struct {
    PyObject_HEAD
    PyObject *db;
    TCLIST *res;
} PDBCursor;

#define PYEJDBTXBEGIN 1
#define PYEJDBTXABORT 2
#define PYEJDBTXCOMMIT 3
#define PYEJDBTXSTATUS 4


#include "pdbcursor.c"
#include "pejdb.c"

PyDoc_STRVAR(ejdb_version_doc, "version() -> str\n\nReturns the version string of the underlying EJDB library.");

static PyObject* ejdb_version(PyObject *module) {
    const char *v = ejdbversion();
    return PyUnicode_FromString(v);
}

/* pyejdb.m_methods */
static PyMethodDef pyejdb_m_methods[] = {
    {"ejdb_version", (PyCFunction) ejdb_version, METH_NOARGS, ejdb_version_doc},
    {NULL} /* Sentinel */
};


#if PY_MAJOR_VERSION >= 3
PyDoc_STRVAR(ejdb_m_doc, "EJDB http://ejdb.org");
#define INITERROR return NULL

/* pyejdb_module */
static PyModuleDef pyejdb_module = {
    PyModuleDef_HEAD_INIT,
    "_pyejdb",          /*m_name*/
    ejdb_m_doc,         /*m_doc*/
    -1,                 /*m_size*/
    pyejdb_m_methods,   /*m_methods*/
};
#else
#define INITERROR return
#endif

PyMODINIT_FUNC init_pyejdb(void) {
    PyObject *pyejdb;
    if (PyType_Ready(&EJDBType) ||
            PyType_Ready(&DBCursorType)
            ) {
        INITERROR;
    }

#if PY_MAJOR_VERSION >= 3
    pyejdb = PyModule_Create(&pyejdb_module);
#else
    pyejdb = Py_InitModule("_pyejdb", pyejdb_m_methods);
#endif

    if (!pyejdb) {
        INITERROR;
    }
    Error = PyErr_NewException("_pyejdb.Error", NULL, NULL);
    Py_XINCREF(Error);
    if (!Error || PyModule_AddObject(pyejdb, "Error", Error)) {
        Py_XDECREF(Error);
        goto fail;
    }

    /* adding types and constants */
    if (
            PyModule_AddType(pyejdb, "EJDB", &EJDBType) ||
            PyModule_AddType(pyejdb, "DBCursor", &DBCursorType) ||

            PyModule_AddIntMacro(pyejdb, JBOREADER) ||
            PyModule_AddIntMacro(pyejdb, JBOWRITER) ||
            PyModule_AddIntMacro(pyejdb, JBOCREAT) ||
            PyModule_AddIntMacro(pyejdb, JBOTRUNC) ||
            PyModule_AddIntMacro(pyejdb, JBONOLCK) ||
            PyModule_AddIntMacro(pyejdb, JBOLCKNB) ||
            PyModule_AddIntMacro(pyejdb, JBOTSYNC) ||

            PyModule_AddIntMacro(pyejdb, JBIDXDROP) ||
            PyModule_AddIntMacro(pyejdb, JBIDXDROPALL) ||
            PyModule_AddIntMacro(pyejdb, JBIDXOP) ||
            PyModule_AddIntMacro(pyejdb, JBIDXREBLD) ||
            PyModule_AddIntMacro(pyejdb, JBIDXNUM) ||
            PyModule_AddIntMacro(pyejdb, JBIDXSTR) ||
            PyModule_AddIntMacro(pyejdb, JBIDXARR) ||
            PyModule_AddIntMacro(pyejdb, JBIDXISTR) ||

            PyModule_AddIntMacro(pyejdb, PYEJDBTXBEGIN) ||
            PyModule_AddIntMacro(pyejdb, PYEJDBTXABORT) ||
            PyModule_AddIntMacro(pyejdb, PYEJDBTXCOMMIT) ||
            PyModule_AddIntMacro(pyejdb, PYEJDBTXSTATUS) ||

            PyModule_AddIntMacro(pyejdb, JBQRYCOUNT)
            ) {

        goto fail;
    }

#if PY_MAJOR_VERSION >= 3
    return pyejdb;
#else
    return;
#endif

fail:
    Py_DECREF(pyejdb);
    INITERROR;
}

PyMODINIT_FUNC PyInit__pyejdb(void) {
    return init_pyejdb();
}



pyejdb/__init__.py
#-*- coding: utf8 -*-

# *************************************************************************************************
#  Python API for EJDB database library http://ejdb.org
#  Copyright (C) 2012-2013 Softmotions Ltd.
#
#  This file is part of EJDB.
#  EJDB is free software; you can redistribute it and/or modify it under the terms of
#  the GNU Lesser General Public License as published by the Free Software Foundation; either
#  version 2.1 of the License or any later version.  EJDB is distributed in the hope
#  that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
#  License for more details.
#  You should have received a copy of the GNU Lesser General Public License along with EJDB;
#  if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
#  Boston, MA 02111-1307 USA.
# *************************************************************************************************

from __future__ import with_statement
from __future__ import division
from __future__ import print_function

import _pyejdb
import sys
from pyejdb.lrucache import lru_cache
from pyejdb import bson
from collections import OrderedDict as odict
from io import StringIO, BytesIO
import re, numbers


PY3 = sys.version_info[0] == 3

__all__ = [

    "EJDB",

    "JBOREADER",
    "JBOWRITER",
    "JBOCREAT",
    "JBOTRUNC",
    "JBONOLCK",
    "JBOLCKNB",
    "JBOTSYNC",
    "DEFAULT_OPEN_MODE",

    "JBQRYCOUNT",

    "check_oid",
    "version",
    "libejdb_version"
]

version_tuple = (1, 0, 16)

def get_version_string():
    return '.'.join(map(str, version_tuple))

version = get_version_string()
libejdb_version = _pyejdb.ejdb_version()

#OID check RE
_oidRE = re.compile("^[0-9a-f]{24}$")

#Open modes
JBOREADER = _pyejdb.JBOREADER
JBOWRITER = _pyejdb.JBOWRITER
JBOCREAT = _pyejdb.JBOCREAT
JBOTRUNC = _pyejdb.JBOTRUNC
JBONOLCK = _pyejdb.JBONOLCK
JBOLCKNB = _pyejdb.JBOLCKNB
JBOTSYNC = _pyejdb.JBOTSYNC
DEFAULT_OPEN_MODE = JBOWRITER | JBOCREAT

#Query flags
JBQRYCOUNT = _pyejdb.JBQRYCOUNT

#Misc
def check_oid(oid):
    if PY3:
        if not isinstance(oid, str) or _oidRE.match(oid) is None:
            raise ValueError("Invalid OID: %s" % oid)
    else:
        tn = type(oid).__name__
        if (tn != "unicode" and tn != "str") or _oidRE.match(oid) is None:
            raise ValueError("Invalid OID: %s" % oid)


class EJDBCursorWrapper(object):
    """ EJDB resultset cursor returned by `EJDB.find()`.
    Sequences and iterator protocols supported.
    """

    def __init__(self, dbcursor):
        self.__pos = 0
        self.__cursor = dbcursor
        self.__len = self.__cursor.length()

    def __del__(self):
        self.__cursor.close()

    def __len__(self):
        return self.__len

    def __getitem__(self, key):
        if isinstance(key, slice):
            return [self[idx] for idx in range(*key.indices(len(self)))]
        elif isinstance(key, numbers.Number):
            if key < 0:
                key += len(self)
            if key >= len(self):
                raise IndexError("The index (%d) is out of range." % key)
            return self.get(key)
        else:
            raise TypeError("Invalid argument type.")

    def __iter__(self):
        return self

    def __next__(self):
        if self.__pos >= self.__len:
            raise StopIteration
        self.__pos += 1
        return self.get(self.__pos - 1)

    def next(self):
        return self.__next__()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def reset(self):
        """ Resets iterator to its initial state.
        """
        self.__pos = 0

    @lru_cache(maxsize=1024)
    def get(self, idx):
        """Return JSON document at the specified position `idx`
        """
        bsdata = self.__cursor.get(idx)
        return bson.parse_bytes_lazy(bsdata) if bsdata is not None else None

    def close(self):
        """ Closes cursor and frees all allocated resources.
        Closed cursor became unusable and its
        methods will raise "Closed cursor" error.
        """
        self.__cursor.close()


class EJDB(object):
    """ EJDB database wrapper class.
    Typical creation:
    >>> ejdb = pyejdb.EJDB("<db file path>", <db open mode>)

    :Parameters:
        - `fpath`: Database file path
        - `mode` (optional): Open mode int mask
            JBOREADER - Open as reader
            JBOWRITER - Open as writer
            JBOCREAT - Create new database if db file not exists
            JBOTRUNC - Truncate database on open
            JBONOLCK - Open without locking
            JBOLCKNB - Lock without blocking
            JBOTSYNC - Synchronize db on every transaction

    Default open mode: DEFAULT_OPEN_MODE = JBOWRITER | JBOCREAT
    """

    def __init__(self, fpath, mode=DEFAULT_OPEN_MODE):
        self.__ejdb = _pyejdb.EJDB()
        self.__ejdb.open(fpath, mode)

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    @property
    def isopen(self):
        return self.__ejdb.isopen()

    def close(self):
        return self.__ejdb.close()

    def sync(self):
        return self.__ejdb.sync()

    def command(self, cmd):
        """ Executes ejdb database command.

          Supported commands:

           1) Exports database collections data. See ejdbexport() method.

             "export" : {
                   "path" : string,                    //Exports database collections data
                   "cnames" : [string array]|null,     //List of collection names to export
                   "mode" : int|null                   //Values: null|`JBJSONEXPORT` See ejdb.h#ejdbexport() method
             }

             Command response:
                {
                   "log" : string,        //Diagnostic log about executing this command
                   "error" : string|null, //ejdb error message
                   "errorCode" : int|0,   //ejdb error code
                }

           2) Imports previously exported collections data into ejdb.

             "import" : {
                   "path" : string                     //The directory path in which data resides
                   "cnames" : [string array]|null,     //List of collection names to import
                   "mode" : int|null                //Values: null| JBIMPORTUPDATE`|`JBIMPORTREPLACE` See ejdb.h#ejdbimport() method
              }

              Command response:
                {
                   "log" : string,        //Diagnostic log about executing this command
                   "error" : string|null, //ejdb error message
                   "errorCode" : int|0,   //ejdb error code
                }

         :Parameters:
            - `cmd`  Command object dictionary

         :Returns:
            Command response object dictionary
        """
        bret = self.__ejdb.command(bson.serialize_to_bytes(cmd))
        if bret is not None:
            return bson.parse_bytes(bret)
        return None

    def save(self, cname, *jsarr, **kwargs):
        """ Save/update specified `dict` documents into collection `cname`.

        Samples:
        >>> ejdb.save('foo', {'foo' : 'bar'})
        >>> ejdb.save('foo', {_id : '511c72ae7922641d00000000', 'foo' : 'bar'}, {'foo' : 'bar2'}, merge=True)
        >>> ejdb.save('foo', *[{'foo' : 'bar'}, {'foo' : 'bar2'}, ...])

        If collection with `cname` does not exists it will be created.
        Each document may have unique identifier (OID) stored in the `_id` property.
        If a saved doc does not have `_id` it will be autogenerated,
            document `dict` object will be updated with this `_id`
            and new document record will be stored.
        To identify and update doc it should contains `_id` dict property.

        :Parameters:
            - `cname` Collection name
            - `*jsarr` Variable arg list with doc to be saved
            -  merge=False (optional) If `True` docs with `_id` will be merged with those stored in db,
               otherwise updated documents will be fully replaced by new instances.
        """
        for doc in jsarr:
            if doc is not None:
                _oid = self.__ejdb.save(cname, bson.serialize_to_bytes(doc), **kwargs)
                if "_id" not in doc:
                    doc["_id"] = _oid

    def load(self, cname, oid):
        """ Loads `dict` documents identified by `oid` from `cname` collection.
        Sample:
        >>> ejdb.load('mycoll', '511c72ae7922641d00000000');
        :Parameters:
            - `cname` Collection name
            - `oid` Document object id (`_id` property)
        :Returns:
            A document `dict` identified by `oid` or `None` if it is not found
        """
        check_oid(oid)
        docbytes = self.__ejdb.load(cname, oid)
        if docbytes is None:
            return None
        return bson.parse_bytes(docbytes)

    def remove(self, cname, oid):
        """ Removes from `cname` collection the document identified by `oid`
        """
        check_oid(oid)
        return self.__ejdb.remove(cname, oid)

    def find(self, cname, qobj=None, *args, **kwargs):
        """ Execute query on collection.

        Sample:
        >>> # Fetch all elements from collection
        >>> ejdb.find("mycoll")
        >>> # Query document with 'foo==bar' condition, include in resulting doc 'foo' and '_id' fields only
        >>> ejdb.find("mycoll", {'foo' : 'bar'}, hints={$fields : {'foo' : 1, '_id' : 1}});

        General format:
            find(<collection name>, <query object>, <OR joined query objects>,..., hints=<Query Hints>)

        :Parameters:
           - `cname` Collection name
           - `qobj` Main query object
           - *args OR joined query object
           - hints={} (optional) Query hints. See explanations below.
           - log=StringIO (optional) StringIO buffer for ejdb query log (debugging mode)

        :Returns:
            Resultset cursor :class:`EJDBCursorWrapper`

        EJDB queries inspired by MongoDB (mongodb.org) and follows same philosophy.
        - Supported queries:
           - Simple matching of String OR Number OR Array value:
               -   {'fpath' : 'val', ...}
           - $not Negate operation.
               -   {'fpath' : {'$not' : val}} //Field not equal to val
               -   {'fpath' : {'$not' : {'$begin' : prefix}}} //Field not begins with val
           - $begin String starts with prefix
               -   {'fpath' : {'$begin' : prefix}}
           - $gt, $gte (>, >=) and $lt, $lte for number types:
               -   {'fpath' : {'$gt' : number}, ...}
           - $bt Between for number types:
               -   {'fpath' : {'$bt' : [num1, num2]}}
           - $in String OR Number OR Array val matches to value in specified array:
               -   {'fpath' : {'$in' : [val1, val2, val3]}}
           - $nin - Not IN
           - $strand String tokens OR String array val matches all tokens in specified array:
               -   {'fpath' : {'$strand' : [val1, val2, val3]}}
           - $stror String tokens OR String array val matches any token in specified array:
               -   {'fpath' : {'$stror' : [val1, val2, val3]}}
           - $exists Field existence matching:
               -   {'fpath' : {'$exists' : true|false}}
           - $icase Case insensitive string matching:
               -    {'fpath' : {'$icase' : 'val1'}} //icase matching
               icase matching with '$in' operation:
               -    {'name' : {'$icase' : {'$in' : ['tHéâtre - театр', 'heLLo WorlD']}}}
               For case insensitive matching you can create special type of string index.
           - $elemMatch The $elemMatch operator matches more than one component within an array element.
               -    { array: { $elemMatch: { value1 : 1, value2 : { $gt: 1 } } } }
               Restriction: only one $elemMatch allowed in context of one array field.
           - $and, $or joining:
               -   {..., $and : [subq1, subq2, ...] }
               -   {..., $or  : [subq1, subq2, ...] }
              Example: {z : 33, $and : [ {$or : [{a : 1}, {b : 2}]}, {$or : [{c : 5}, {d : 7}]} ] }

        - Queries can be used to update records:

           - $set Field set operation.
               - {.., '$set' : {'field1' : val1, 'fieldN' : valN}}
           - $upsert Atomic upsert. If matching records are found it will be '$set' operation,
                     otherwise new record will be inserted with fields specified by argment object.
               - {.., '$upsert' : {'field1' : val1, 'fieldN' : valN}}
           - $inc Increment operation. Only number types are supported.
               - {.., '$inc' : {'field1' : number, ...,  'field1' : number}
           - $dropall In-place record removal operation.
               - {.., '$dropall' : True}
           - $addToSet Atomically adds value to the array only if its not in the array already.
                       If containing array is missing it will be created.
               - {.., '$addToSet' : {'fpath' : val1, 'fpathN' : valN, ...}}
           - $addToSetAll Batch version if $addToSet
               - {.., '$addToSetAll' : {'fpath' : [array of values to add], ...}}
           - $pull Atomically removes all occurrences of value from field, if field is an array.
               - {.., '$pull' : {'fpath' : val1, 'fpathN' : valN, ...}}
           - $pullAll Batch version of $pull
               - {.., '$pullAll' : {'fpath' : [array of values to remove], ...}}

        - Collection joins supported in the following form:
           - {..., $do : {fpath : {$join : 'collectionname'}} }
            Where 'fpath' value points to object's OIDs from 'collectionname'. Its value
            can be OID, string representation of OID or array of this pointers.

        .. NOTE:: It is better to execute update queries with `$onlycount=true` hint flag
               or use the special `update()` method to avoid unnecessarily data fetching.

        .. NOTE:: Negate operations: $not and $nin not using indexes
                 so they can be slow in comparison to other matching operations.

        .. NOTE:: Only one index can be used in search query operation.


        QUERY HINTS (specified by `hints=` argument):
           - $max Maximum number in the result set
           - $skip Number of skipped results in the result set
           - $orderby Sorting order of query fields.
           - $onlycount true|false If `true` only count of matching records will be returned
                                   without placing records in result set.
           - $fields Set subset of fetched fields.
               If field presented in $orderby clause it will be forced to include in resulting records.
               Example:
               hints:    {
                   "$orderby" : [ //ORDER BY field1 ASC, field2 DESC
                       ("field1", 1),
                       ("field2", -1)
                   ],
                   "$fields" : { //SELECT ONLY {_id, field1, field2}
                       "field1" : 1,
                       "field2" : 1
                   }
               }

        """
        if not qobj: qobj = {}
        qobj = bson.serialize_to_bytes(qobj)
        hints = bson.serialize_to_bytes(self.__preprocessQHints(kwargs.get("hints", {})))
        orarr = [bson.serialize_to_bytes(x) for x in args]
        qflags = kwargs.get("qflags", 0)
        log = kwargs.get("log")
        log = log if isinstance(log, StringIO) or isinstance(log, BytesIO) else None
        cursor = self.__ejdb.find(cname, qobj, orarr, hints, qflags, log)
        return cursor if isinstance(cursor, numbers.Number) else EJDBCursorWrapper(cursor)

    def findOne(self, cname, qobj=None, *args, **kwargs):
        """ Same as `#find()` but retrieves only one matching JSON object.
        """
        hints = self.__preprocessQHints(kwargs.get("hints", {}))
        hints["$max"] = 1
        kwargs["hints"] = hints
        with self.find(cname, qobj, *args, **kwargs) as res:
            return res[0] if len(res) > 0 else None

    def update(self, cname, qobj=None, *args, **kwargs):
        """ Convenient method to execute update queries.
        :Returns:
            Count of updated objects
        """
        qflags = kwargs.get("qflags", 0)
        kwargs["qflags"] = qflags | JBQRYCOUNT
        return self.find(cname, qobj, *args, **kwargs)

    def count(self, cname, qobj=None, *args, **kwargs):
        """ Counts matched documents.
        :Returns:
            Count of matched objects
        """
        qflags = kwargs.get("qflags", 0)
        kwargs["qflags"] = qflags | JBQRYCOUNT
        return self.find(cname, qobj, *args, **kwargs)

    def dbmeta(self):
        """ Retrieve metainfo object describing database structure.
        """
        return self.__ejdb.dbmeta()

    def begintx(self, cname):
        """ Begin collection transaction.
        """
        return self.__ejdb.txcontrol(cname, _pyejdb.PYEJDBTXBEGIN)

    def commitx(self, cname):
        """ Commit collection transaction.
        """
        return self.__ejdb.txcontrol(cname, _pyejdb.PYEJDBTXCOMMIT)

    def abortx(self, cname):
        """ Abort collection transaction.
        """
        return self.__ejdb.txcontrol(cname, _pyejdb.PYEJDBTXABORT)

    def isactivetx(self, cname):
        """ Is collection transaction active
        """
        return self.__ejdb.txcontrol(cname, _pyejdb.PYEJDBTXSTATUS)

    def ensureCollection(self, cname, **kwargs):
        """ Automatically creates new collection if it does not exists.
        Collection options `copts`
        are applied only for newly created collection.
        for existing collections `copts` takes no effect.
        Collection options (kwargs):
            {
               "cachedrecords" : Max number of cached records in shared memory segment. Default: 0
               "records" : Estimated number of records in this collection. Default: 65535.
               "large" : Specifies that the size of the database can be larger than 2GB. Default: false
               "compressed" : If true collection records will be compressed with DEFLATE compression. Default: false.
            }
        """
        return self.__ejdb.ensureCollection(cname, **kwargs)

    def dropCollection(self, cname, **kwargs):
        """ Removes database collection.
        """
        return self.__ejdb.dropCollection(cname, **kwargs)

    def dropIndexes(self, cname, path):
        """ Drops indexes of all types for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXDROPALL)

    def optimizeIndexes(self, cname, path):
        """ Optimize indexes of all types for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXOP)

    def ensureStringIndex(self, cname, path):
        """ Ensure index presence of String type for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXSTR)

    def rebuildStringIndex(self, cname, path):
        """ Rebuild index of String type for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXSTR | _pyejdb.JBIDXREBLD)

    def dropStringIndex(self, cname, path):
        """ Drop index of String type for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXSTR | _pyejdb.JBIDXDROP)

    def ensureIStringIndex(self, cname, path):
        """ Ensure case insensitive String index for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXISTR)

    def rebuildIStringIndex(self, cname, path):
        """Rebuild case insensitive String index for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXISTR | _pyejdb.JBIDXREBLD)

    def dropIStringIndex(self, cname, path):
        """Drop case insensitive String index for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXISTR | _pyejdb.JBIDXDROP)

    def ensureNumberIndex(self, cname, path):
        """Ensure index presence of Number type for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXNUM)

    def rebuildNumberIndex(self, cname, path):
        """Rebuild index of Number type for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXNUM | _pyejdb.JBIDXREBLD)

    def dropNumberIndex(self, cname, path):
        """Drop index of Number type for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXNUM | _pyejdb.JBIDXDROP)

    def ensureArrayIndex(self, cname, path):
        """Ensure index presence of Array type for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXARR)

    def rebuildArrayIndex(self, cname, path):
        """Rebuild index of Array type for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXARR | _pyejdb.JBIDXREBLD)

    def dropArrayIndex(self, cname, path):
        """Drop index of Array type for JSON field path.
        """
        return self.__ejdb.setIndex(cname, path, _pyejdb.JBIDXARR | _pyejdb.JBIDXDROP)

    def __preprocessQHints(self, hints):
        val = hints.get("$orderby")
        if isinstance(val, list):
            hints["$orderby"] = odict(val)
        return hints
